app/controllers/api/v1/time_sheet_entries_controller.rb

```
@@ -0,0 +1,34 @@
+class Api::V1::TimeSheetEntriesController < ApplicationController
+  before_action :authenticate_user!
+  before_action :set_resources, only: :create
+
+  def create
+    time_sheet_entry = TimeSheetEntry.new(time_sheet_entry_params)
+
+    if time_sheet_entry.save
+      render json: TimeSheetEntrySerializer.new(time_sheet_entry).serializable_hash, status: :created
+    else
+      render json: { errors: time_sheet_entry.errors.full_messages }, status: :unprocessable_entity
+    end
+  end
+
+  private
+
+  def time_sheet_entry_params
+    params.require(:time_sheet_entry).permit(:user_id, :project_id, :time_period_id, :total_hours)
+  end
+
+  def set_resources
+    unless User.exists?(params[:time_sheet_entry][:user_id])
+      render json: { error: 'User not found' }, status: :not_found and return
+    end
+
+    unless Project.exists?(params[:time_sheet_entry][:project_id])
+      render json: { error: 'Project not found' }, status: :not_found and return
+    end
+
+    return if TimePeriod.exists?(params[:time_sheet_entry][:time_period_id])
+
+    render json: { error: 'Time Period not found' }, status: :not_found and return
+  end
```

I noticed that the set_resources method has some duplicated validation logic. To improve readability and avoid repetition, we could extract it into a reusable validate_resource method and uses throw(:abort) instead of and return for cleaner flow control (https://guides.rubyonrails.org/active_record_callbacks.html#:~:text=this%20author%22)-,throw(%3Aabort),-end%0A%20%20%20%20end%0Aend) 

Here’s how it would look:

```
def validate_resource(model, id_param, error_message)
  id_value = params.dig(:time_sheet_entry, id_param)
  unless id_value && model.exists?(id_value)
    render json: { error: error_message }, status: :not_found
    throw(:abort)
  end
end

def set_resources
  validate_resource(User, :user_id, 'User not found')
  validate_resource(Project, :project_id, 'Project not found')
  validate_resource(TimePeriod, :time_period_id, 'Time Period not found')
end
```

---

Done